### PR TITLE
Refactor count properties to custom managers

### LINKED ISSUE #622: Refactor some model properties to managers
Properties such as `Workshop.solution_count` are an easy way to introduce N+1 errors in the ORM. Instead, we should remove the property and define a custom ObjectManager with a method `with_counts` that adds the required data at the query level.

### PR DESCRIPTION
**Pull request by krzys-h**
This reduces the SQL queries required to load the 'Workshops' view significantly, and helps prevent ORM N+1 problems in the future

Closes #622

- - -
This change is [Reviewable](https://reviewable.io/reviews/warsztatywww/aplikacjawww/623)

### CODE REVIEW

File: wwwapp/views.py [hunk 1]
```diff
@@ -615,47 +618,25 @@ def participants_view(request, year=None):
         if year:
             assert camp_participation is not None
             for wp in camp_participation.workshop_participation.all():
-                if wp.workshop.is_qualifying:
-                    if not wp.workshop.solution_uploads_enabled or hasattr(wp, 'solution'):
-                        people[participant.id]['to_be_checked_solution_count'] += 1
-
-                    if wp.qualification_result:
-                        people[participant.id]['points'] += float(wp.result_in_percent())
```
Comment (krzys-h) [inline wwwapp/views.py:623]: I just realized I forgot about the points... oops, will fix later

Comment (krzys-h) [reply wwwapp/views.py:623]: Done

**Review by cytadela8 — COMMENTED** [review]
I made progress in reviewing this :smiley: 

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 6 of 6 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/warsztatywww/aplikacjawww/623#-:-NB2eaDrCOOLdtafKeg2:by9tbek)* status: 5 of 8 files reviewed, all discussions resolved (waiting on @cytadela8)

___
*[`wwwapp/models.py` line 394 at r3](https://reviewable.io/reviews/warsztatywww/aplikacjawww/623#-NB2Y5ou4FtdT7FtEOJn:-NB2Y5ou4FtdT7FtEOJo:b4fzbcy) ([raw file](https://github.com/warsztatywww/aplikacjawww/blob/cacdfa0b6f63207ad184f64921f3bc8ca620fc95/wwwapp/models.py#L394)):*
> ```Python
>                 default=None
>             ),
>         )
> ```

BTW, reading this makes me hate ORMs...



_Suggestion:_
```Python
            registered_count=Count('participants'),
            solution_count=Case(
                When(is_qualifying=True, solution_uploads_enabled=True, then=Count('participants__solution')),
                default=None
            ),
            checked_solution_count=Case(
                When(is_qualifying=True, solution_uploads_enabled=True, then=Count('participants__qualification_result', filter=Q(participants__solution__isnull=False))),
                When(is_qualifying=True, solution_uploads_enabled=False, then=Count('participants__qualification_result')),
                default=None,
            ),Tłumaczenie strony
Learn how to fix: "SSH Authentication Refused: Bad Ownership or Modes for Directory" error when trying to set up public key authenticated automatic logins.

Isilon: OneFS: Error-"Authentication refused: bad ownership or ...
https://www.dell.com › kbdoc › en-us
Tłumaczenie strony
22 kwi 2021 — The error, "Authentication refused: bad ownership or modes for directory," typically means that the user's home or ~/.ssh directories have group ...

SSH Authentication Refused: Bad Ownership or Modes for ...
http://miro.borodziuk.eu › index.php › 2020/01/21 › ss...
21 sty 2020 — Finally we're getting somewhere – bad ownership or modes for directory /home/dave/.ssh. SSH doesn't like it if your home or ~/.ssh ...

Fix: SSH Authentication Refused: Bad Ownership or Modes for ...
            qualified_count=Case(
                When(is_qualifying=True, qualification_threshold__isnull=False, then=Count('participants', filter=Q(participants__qualification_result__gte=F('qualification_threshold')))),
                default=None
            ),
        )
```

**Review by cytadela8 — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r6, 3 of 3 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/warsztatywww/aplikacjawww/623)* status: all files reviewed, 1 unresolved discussion (waiting on @krzys-h)

___
*[`wwwapp/views.py` line 605 at r7](https://reviewable.io/reviews/warsztatywww/aplikacjawww/623#-NB3893cDLqM4cbWXzEd:-NB3893cDLqM4cbWXzEe:b-ulb9og) ([raw file](https://github.com/warsztatywww/aplikacjawww/blob/47a3bb0698689602bf8679ef09fbfdb4e9aaad87/wwwapp/views.py#L605)):*
> ```Python
>             'to_be_checked_solution_count': camp_participation.to_be_checked_solution_count if camp_participation else 0,
>             'accepted_workshop_count': camp_participation.accepted_workshop_count if camp_participation else 0,
>             'checked_solution_percentage': camp_participation.checked_solution_percentage if camp_participation else -1,
> ```

`-1` ?

**Review by krzys-h — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/warsztatywww/aplikacjawww/623)* status: all files reviewed, 1 unresolved discussion (waiting on @cytadela8)

___
*[`wwwapp/views.py` line 605 at r7](https://reviewable.io/reviews/warsztatywww/aplikacjawww/623#-NB3893cDLqM4cbWXzEd:-NB3FpES7jsnrylSjDdV:bo1ewke) ([raw file](https://github.com/warsztatywww/aplikacjawww/blob/47a3bb0698689602bf8679ef09fbfdb4e9aaad87/wwwapp/views.py#L605)):*
<details><summary><i>Previously, cytadela8 (Artur Puzio) wrote…</i></summary><blockquote>

`-1` ?
</blockquote></details>

This is how `checked_solution_percentage` always worked internally - the point is to make "not submitted anything" appear lower than "submitted something, but 0 of the solutions are checked" when sorting

The value here is probably never used though, as `camp_participation` can be None only when rendering the all people view which doesn't show this column at all

**Review by cytadela8 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/warsztatywww/aplikacjawww/623)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**